h/static/scripts/build-thread.js

```diff
@@ -0,0 +1,295 @@
+'use strict';
+
+/** Default state for new threads, before applying filters etc. */
+var DEFAULT_THREAD_STATE = {
+  /** The Annotation which is displayed by this thread. */
+  annotation: undefined,
+  /** The parent Thread */
+  parent: undefined,
+  /** True if replies to this annotation are hidden. */
+  collapsed: false,
+  /** True if this annotation matches the current filters. */
+  visible: true,
+  /** Replies to this annotation. */
+  children: [],
+  /**
+    * The total number of children of this annotation,
+    * including any which have been hidden by filters.
+    */
+  totalChildren: 0,
+};
+
+/**
+ * Returns a persistent identifier for an Annotation.
+ * If the Annotation has been created on the server, it will have
+ * an ID assigned, otherwise we fall back to the local-only '$$tag'
+ * property.
+ */
+function id(annotation) {
+  return annotation.id || annotation.$$tag;
+}
+
+/**
+ * Creates a thread of annotations from a list of annotations.
```

Comment:
I certainly agree that there is a design discussion to be had here. For now I think there are probably many improvements that can be made to the code without changing existing behaviour. One question: I'm not sure how the new code works when intermediate replies are deleted, but here's a scenario raised by the example you gave above:

```
annotation
|
` reply 1
  |
   ` reply 2
```

What happens if reply 1 is deleted? The JWZ algorithm allows us to infer the presence of the missing message and display that information to the user. While we have explicit threaded conversations I think it's important that we don't change to something that essentially lies about what happened. For example, I think producing the rendered output

```
annotation
|
` reply 2
```

would be problematic.